src/main/java/com/lovetropics/minigames/common/util/world/gamedata/GameDataAccessor.java

```diff
@@ -0,0 +1,111 @@
+
+    static final SuggestionProvider<CommandSourceStack> SUGGEST_GAMEDATA = ((context, builder) -> SharedSuggestionProvider.suggestResource(
+            getGameDataStorage(context).playerData.keySet(), builder
+    ));
+
+    public static final Function<String, DataCommands.DataProvider> PROVIDER = (str) -> new DataCommands.DataProvider() {
+        @Override
+        public DataAccessor access(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
+            return new GameDataAccessor(context.getSource().getLevel(), getGameDataStorage(context),
+                    ResourceLocationArgument.getId(context, str), EntityArgument.getPlayer(context, "player").getUUID());
+        }
+
+        @Override
+        public ArgumentBuilder<CommandSourceStack, ?> wrap(ArgumentBuilder<CommandSourceStack, ?> builder, Function<ArgumentBuilder<CommandSourceStack, ?>, ArgumentBuilder<CommandSourceStack, ?>> action) {
+            return builder.then(
+                    Commands.literal("gamedata")
+                            .then(Commands.argument(str, ResourceLocationArgument.id()).suggests(GameDataAccessor.SUGGEST_GAMEDATA)
+                                    .then(action.apply(Commands.argument("player", EntityArgument.player()))))
+            );
+        }
+    };
+
+    @NotNull
+    private static GameDataStorage getGameDataStorage(CommandContext<CommandSourceStack> context) {
+        return GameDataStorage.get(context.getSource().getLevel());
+    }
+    private final Level level;
+    private final GameDataStorage gameDataStorage;
+    private final ResourceLocation id;
+    private final UUID player;
+
+    public GameDataAccessor(Level level, GameDataStorage gameDataStorage, ResourceLocation id, UUID player) {
+        this.level = level;
+        this.gameDataStorage = gameDataStorage;
+        this.id = id;
+        this.player = player;
+    }
+
+    @Override
+    public void setData(CompoundTag other) throws CommandSyntaxException {
+        gameDataStorage.set(id, player, other);
+    }
+
+    @Override
+    public CompoundTag getData() throws CommandSyntaxException {
+        return gameDataStorage.get(id, player);
+    }
+
+    @Override
+    public Component getModifiedSuccess() {
+        String playerName = getPlayerName();
+        return STORAGE_MODIFIED.apply(playerName, Component.translationArg(this.id));
+    }
+
+    @Override
+    public Component getPrintSuccess(Tag nbt) {
+        String playerName = getPlayerName();
+        return STORAGE_QUERY.apply(playerName, Component.translationArg(this.id), NbtUtils.toPrettyComponent(nbt));
+    }
+
+    @Override
+    public Component getPrintSuccess(NbtPathArgument.NbtPath path, double scale, int value) {
+        String playerName = getPlayerName();
+        return STORAGE_GET.apply(path.asString(), Component.translationArg(this.id), playerName, String.format(Locale.ROOT, "%.2f", scale), value);
+    }
+
+    private String getPlayerName() {
```

Comment:
Nit: we can avoid these lookups if we pass the GameProfile directly into here :)